`zonal_statistics`: expand bbox intersection to align with grid

emlys opened the pull request:
This PR updates `zonal_statistics` to make sure that the bounding box used is aligned with the input raster stack. Coincidentally fixes #332.

The original bounding box is the intersection between the input raster(s) and vector. This box doesn't necessarily align with the input raster's grid. Having the intermediate rasters' grid shifted by a fraction of a pixel can change the results, because it changes which pixels' centerpoints fall within a polygon.

Keeping the grid aligned with the input will keep results consistent, independent of the disjoint polygon sets' bounding boxes. This also resolves most of the places where invest test results were changed by pygeoprocessing 2.4.1.

emlys committed:
expand bbox intersection to align with grid

emlys committed:
align_bbox: adjust and test for positive and negative pixel sizes

emlys committed:
clean up

emlys committed:
clean up

emlys commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1357,6 +1357,44 @@ def interpolate_points(
+        y_origin + abs(pixel_height) * math.floor(
+            (bbox[1] - y_origin) / pixel_height * numpy.sign(pixel_height)),
+        x_origin + abs(pixel_width) * math.ceil(
+            (bbox[2] - x_origin) / pixel_width * numpy.sign(pixel_width)),
```
It took me a long time to figure out these equations and I'm still not sure I understand them intuitively. Happy to discuss

emlys replied:
This would be interesting as a tough interview question

phargogh replied:
I agree, this adjustment would be a neat puzzle-style interview question!  Relevant to the domain, useful and also a little bit obscure.

emlys committed:
correct zonal stats test

emlys committed:
clean up

emlys commented on tests/test_geoprocessing.py:
```diff
@@ -1044,8 +1044,8 @@ def test_zonal_stats_all_outside(self):
 
-        x_pos = origin_x + (n+.99)
-        y_pos = origin_y - (n+.99)
+        x_pos = origin_x + (n-.01)
```
This test was failing after the changes to zonal stats in this PR, and I'm not sure why it was passing in the first place. The original polygon is in purple and the new polygon is in brown. According to a comment above, this is meant to test "a polygon that barely intersects the upper left hand corner but is nodata." The original polygon actually intersects 3 valid pixels, with a sum result of `254 + 239 + 238 = 731`, not 0. I moved it up and left by one pixel to be in line with the intent of the test.
<img width="500" alt="Screenshot 2023-09-19 at 11 03 46" src="https://github.com/natcap/pygeoprocessing/assets/43770515/e50ca862-c692-413a-b154-0858cca46e3e">
.

phargogh replied:
Sounds great, and thanks for the graphic!

phargogh commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1357,6 +1357,44 @@ def interpolate_points(
+def align_bbox(geotransform, bbox):
+    """Pad a bounding box so that it aligns with the grid of a given geotransform.
+
+    Ignores row and column rotation.
```
If the user does provide rotation, could we log a warning or something like that?  I'm not entirely sure it's a failure state to provide rotation here and the output bbox will be close-ish to what it should be anyways, so it'd be great if we could at least have something appear in the logfile to help us debug if it were to happen.

emlys replied:
Sure!

phargogh commented on src/pygeoprocessing/geoprocessing.py:
```diff
@@ -1357,6 +1357,44 @@ def interpolate_points(
+    x_origin, pixel_width, _, y_origin, _, pixel_height = geotransform
+    if pixel_width == 0 or pixel_height == 0:
+        raise ValueError('Pixel width and height must not be 0')
+    if bbox[2] < bbox[0] or bbox[3] < bbox[1]:
+        raise ValueError('bbox must be in order [minX, minY, maxX, maxY]')
```
This is helpful!  I've been bitten by the ordering of bbox coordinates many times.

phargogh reviewed (changes requested):
This looks great @emlys !  Could you add a note to HISTORY, since we're exposing a new function at the top module level?

emlys committed:
zonal_stats warp call: avoid VRT, specify resolution

emlys committed:
align_bbox: log a warning if row or column rotation is provided

emlys committed:
Merge branch 'main' into task/zonal-stats-align-rasterized-fids

emlys committed:
add history note

emlys commented:
@phargogh I added a log warning if the geotransform includes row/column rotation, and also fixed an edge case that came up when running the invest tests.

emlys commented:
After looking more at the similar functionality in `create_raster_from_bounding_box`, I'm not seeing enough overlap that would make sense to factor out, but let me know if you disagree!

phargogh commented:
Yeah, let's leave them separate for now.  I think there might be a way for us to merge the functionality (especially now that we have 3 semi-related use cases!) but we should design the solution and implement/refactor in another PR.  I just created #348 for this, but if it does make sense to do, I think that'd make sense to do in a separate PR!
